proposals/lip-0039.md

```diff
@@ -302,273 +328,365 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
-        else if d == 1:
-            update q.hash to branchHash(siblingHash, q.hash)
-
-        remove the first digit of q.binaryBitmap
-        sortedQueries=sortByBitmapAndKey(sortedQueries)
-        sortedQueries=filterQueries(sortedQueries)
-
-    throw error
+    # Finally calculate the root from the provided proof and check that it matches the input Merkle root.
+    return calculate_root(proof.sibling_hashes, filtered_queries) == root
 ```
 
-```java
-function verify(queryKeys, proof, merkleRoot):
-    // check that the proof is valid with respect to queryKeys
-    if length of queryKeys != length of proof.queries:
-        return false
-    for each query q in proof.queries:
-        let k be the corresponding key in queryKeys (the element of queryKeys with same index)
-        // q is an inclusion proof for k or a default empty node
-        if q.key == k:
-            continue
-        // q is an inclusion proof for another leaf node
-        let binaryResponseBitmap = binaryExpansion(q.bitmap) without leading zeros
-        let binaryResponseKey=binaryExpansion(q.key)
-        let binaryQueryKey=binaryExpansion(k)
-        let sharedPrefix = the common prefix of binaryResponseKey and binaryQueryKey
-        if length of binaryResponseBitmap > length of sharedPrefix:
-            // q does not give an non-inclusion proof for k
-            return false
-
-    // calculate the new Merkle root from the proof
-    let calculatedRoot = calculateRoot(proof.siblingHashes, proof.queries)
-    return calculatedRoot == merkleRoot
+```python
+def calculate_root(sibling_hashes: list[bytes], queries: list[Query]) -> bytes:
+    sorted_queries = sorted(queries, key=lambda q: (-q.height, q.key))
+
+    while len(sorted_queries) > 0:
+        q = sorted_queries.pop(0)
+
+        # We reached the top of the tree, return merkle root
+        if q.height == 0:
+            # Check that no extra bits where appended to the bitmap.
+            assert len(q.binary_bitmap) == 0
+            assert len(sibling_hashes) == 0
+            return q.hash
+
+        # we distinguish three cases for the sibling hash:
+        # 1. sibling is next element of sorted_queries
+        if len(sorted_queries) > 0 and are_sibling_queries(q, sorted_queries[0]):
+            sibling_hash = sorted_queries[0].hash
+            # We are merging two branches.
+            # Check that the bitmap at the merging point is consistent with the nodes type.
+            if sorted_queries[0].hash == EMPTY_HASH:
+                assert q.binary_bitmap[0] == "0"
+            else:
+                assert q.binary_bitmap[0] == "1"
+            if q.hash == EMPTY_HASH:
+                assert sorted_queries[0].binary_bitmap[0] == "0"
+            else:
+                assert sorted_queries[0].binary_bitmap[0] == "1"
+
+            # Check that the bitmap coincide from the merging point up.
+            assert len(q.binary_bitmap) == len(sorted_queries[0].binary_bitmap)
```

Comment:
So the length of the `binary_bitmap` is already checked as `are_sibling_queries` returned `True`. 

```
def are_sibling_queries(q1: Query, q2: Query) -> bool:
    if len(q1.binary_bitmap) != len(q2.binary_bitmap):
        return False

    if q1.binary_key[:q1.height - 1] != q2.binary_key[:q.height - 1]:
        return False

    return q1.binary_key[q1.height - 1] != q.binary_key[q1.height - 1]

```

Reply:
removed unnecessary check